### PR TITLE
[2.7] [Config] [Console] [DependencyInjection] [DomCrawler] [Form] [HttpKernel] [OptionsResolver] [PropertyAccess] [Security] [Translation] [Yaml] static code analysis, code cleanup

### PR DESCRIPTION
**Pull request by kalessil**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

Following is included (no functional changes, found with <a href="https://plugins.jetbrains.com/plugin/7622?pr=phpStorm">Php Inspections (EA Extended)</a>):
- Fixed couple of concatenation as array index warnings
- Slow alter in foreach inspection selectively fixes
- Unset calls can be merged inspection fixes
- array_search in in_array context inspection fixes
- Is null usage inspection fixes
- Prefixed increment/decrement inspection fixes
- Elvis operator can be used inspection fixes
- Alias functions usage inspection fixes
- Ternary operator simplification inspection fixes
- _ _ DIR _ _ equivalent inspection fixes

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 1]
```diff
@@ -109,12 +109,12 @@ private function updateReferences($container, $currentId, $newId)
      */
     private function updateArgumentReferences(array $arguments, $currentId, $newId)
     {
-        foreach ($arguments as $k => $argument) {
+        foreach ($arguments as $k => &$argument) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: :-1: for using refs here and below: it leaves the var as ref outside of the loop, which is error prone. It also favors mem copies because of the way php works internally and often led to perf pb.

Comment (kalessil) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: @nicolas-grekas reverting these changes.

**Comment by kalessil** [comment]
@nicolas-grekas since references in foreach are not accepted, it might be an good idea to check couple of classes:

<code>
    foreach ($value as &$v) // Component/DependencyInjection/ContainerBuilder.php:1007
</code>

<code>
    foreach ($namespacedCommands as &$commands) { // Component/Console/Descriptor/ApplicationDescription.php:149
</code>

<code>
    foreach ($v as $gk => &$gv) { // Component/VarDumper/Cloner/VarCloner.php:131
</code>

<code>
    foreach ($trace as &$t) { // Component/VarDumper/Caster/ExceptionCaster.php:121
</code>

<code>
    foreach ($result as &$entry) { // Component/Form/Extension/Core/DataTransformer/DateTimeToArrayTransformer.php:98
</code>

<code>
   foreach ($messages as &$message) { // Component/Console/Helper/FormatterHelper.php:66
</code>

<code>
   foreach ($trace as &$frame) { // Component/Debug/Exception/FatalErrorException.php:49
</code>

File: src/Symfony/Component/Translation/Tests/TranslatorTest.php [hunk 2]
```diff
@@ -479,7 +479,7 @@ public function testTransChoiceFallbackWithNoTranslation()
     public function testGetMessages($resources, $locale, $expected)
     {
         $locales = array_keys($resources);
-        $_locale = !is_null($locale) ? $locale : reset($locales);
+        $_locale = (null !== $locale) ? $locale : reset($locales);
```
Comment (iltar) [inline src/Symfony/Component/Translation/Tests/TranslatorTest.php]: The parenthesis around are not required in this case

Comment (kalessil) [inline src/Symfony/Component/Translation/Tests/TranslatorTest.php]: @iltar : done

File: src/Symfony/Component/Form/AbstractRendererEngine.php [hunk 3]
```diff
@@ -67,8 +67,10 @@ public function setTheme(FormView $view, $themes)
         // Unset instead of resetting to an empty array, in order to allow
         // implementations (like TwigRendererEngine) to check whether $cacheKey
         // is set at all.
-        unset($this->resources[$cacheKey]);
-        unset($this->resourceHierarchyLevels[$cacheKey]);
+        unset(
+            $this->resources[$cacheKey],
+            $this->resourceHierarchyLevels[$cacheKey]
+        );
```
Comment (fabpot) [inline src/Symfony/Component/Form/AbstractRendererEngine.php]: should be on one line.

File: src/Symfony/Component/Form/FormBuilder.php [hunk 4]
```diff
@@ -138,8 +138,10 @@ public function remove($name)
             throw new BadMethodCallException('FormBuilder methods cannot be accessed anymore once the builder is turned into a FormConfigInterface instance.');
         }
 
-        unset($this->unresolvedChildren[$name]);
-        unset($this->children[$name]);
+        unset(
+            $this->unresolvedChildren[$name],
+            $this->children[$name]
+        );
```
Comment (fabpot) [inline src/Symfony/Component/Form/FormBuilder.php]: should be on one line as well.

File: src/Symfony/Component/Form/FormRenderer.php [hunk 5]
```diff
@@ -296,8 +296,10 @@ public function searchAndRenderBlock(FormView $view, $blockNameSuffix, array $va
         // Clear the caches if they were filled for the first time within
         // this function call
         if ($hierarchyInit) {
-            unset($this->blockNameHierarchyMap[$viewAndSuffixCacheKey]);
-            unset($this->hierarchyLevelMap[$viewAndSuffixCacheKey]);
+            unset(
+                $this->blockNameHierarchyMap[$viewAndSuffixCacheKey],
+                $this->hierarchyLevelMap[$viewAndSuffixCacheKey]
+            );
```
Comment (fabpot) [inline src/Symfony/Component/Form/FormRenderer.php]: same as above.

File: src/Symfony/Component/HttpKernel/EventListener/RouterListener.php [hunk 6]
```diff
@@ -147,8 +147,10 @@ public function onKernelRequest(GetResponseEvent $event)
             }
 
             $request->attributes->add($parameters);
-            unset($parameters['_route']);
-            unset($parameters['_controller']);
+            unset(
+                $parameters['_route'],
+                $parameters['_controller']
+            );
```
Comment (fabpot) [inline src/Symfony/Component/HttpKernel/EventListener/RouterListener.php]: same as above

**Comment by fabpot** [comment]
This should be done on the 2.3 branch

**Comment by kalessil** [comment]
@fabpot : roger that. Let me try re-basing corrected version on 2.3 first.

File: src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php [hunk 7]
```diff
@@ -260,7 +260,7 @@ public function testAddGetCompilerPass()
         $builder->setResourceTracking(false);
         $builderCompilerPasses = $builder->getCompiler()->getPassConfig()->getPasses();
         $builder->addCompilerPass($this->getMock('Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface'));
-        $this->assertEquals(sizeof($builderCompilerPasses) + 1, sizeof($builder->getCompiler()->getPassConfig()->getPasses()));
+        $this->assertEquals(count($builderCompilerPasses) + 1, count($builder->getCompiler()->getPassConfig()->getPasses()));
```
Comment (xabbuh) [inline src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php]: What about using `assertCount()` here instead?

File: src/Symfony/Component/PropertyAccess/PropertyPathBuilder.php [hunk 8]
```diff
@@ -264,8 +264,10 @@ private function resize($offset, $cutLength, $insertionLength)
 
             // All remaining elements should be removed
             for (; $i < $length; ++$i) {
-                unset($this->elements[$i]);
-                unset($this->isIndex[$i]);
+                unset(
+                    $this->elements[$i],
+                    $this->isIndex[$i]
+                );
```
Comment (xabbuh) [inline src/Symfony/Component/PropertyAccess/PropertyPathBuilder.php]: Could be on one line too.

File: src/Symfony/Component/Security/Acl/Dbal/AclProvider.php [hunk 9]
```diff
@@ -192,7 +193,8 @@ public function findAcls(array $oids, array $sids = array())
                         $this->cache->putInCache($loadedAcl);
                     }
 
-                    if (isset($oidLookup[$loadedOid->getIdentifier().$loadedOid->getType()])) {
+                    $strOidLookupKey = $loadedOid->getIdentifier() . $loadedOid->getType();
```
Comment (xabbuh) [inline src/Symfony/Component/Security/Acl/Dbal/AclProvider.php]: Symfony doesn't use spaces around the string concatenation operator.

File: src/Symfony/Component/Security/Acl/Dbal/AclProvider.php [hunk 10]
```diff
@@ -559,10 +561,11 @@ private function hydrateObjectIdentities(Statement $stmt, array $oidLookup, arra
                 // attach ACL to the result set; even though we do not enforce that every
                 // object identity has only one instance, we must make sure to maintain
                 // referential equality with the oids passed to findAcls()
-                if (!isset($oidCache[$objectIdentifier.$classType])) {
-                    $oidCache[$objectIdentifier.$classType] = $acl->getObjectIdentity();
+                $strOidCacheKey = $objectIdentifier . $classType;
```
Comment (xabbuh) [inline src/Symfony/Component/Security/Acl/Dbal/AclProvider.php]: same here

File: src/Symfony/Component/Security/Acl/Dbal/AclProvider.php [hunk 11]
```diff
@@ -173,7 +173,8 @@ public function findAcls(array $oids, array $sids = array())
             }
 
             // Is it time to load the current batch?
-            if ((self::MAX_BATCH_SIZE === count($currentBatch) || ($i + 1) === $c) && count($currentBatch) > 0) {
+            $intCurrentBatchesCount = count($currentBatch);
```
Comment (xabbuh) [inline src/Symfony/Component/Security/Acl/Dbal/AclProvider.php]: Should we really start using type prefixes in variable names?

File: src/Symfony/Component/Security/Acl/Dbal/AclProvider.php (hunk 9 above)
Comment (xabbuh) [inline src/Symfony/Component/Security/Acl/Dbal/AclProvider.php]: By the way, what is the reason for this change at all? The resulting string is only used once.

File: src/Symfony/Component/DomCrawler/Tests/CrawlerTest.php [hunk 12]
```diff
@@ -327,7 +327,7 @@ public function testReduce()
     {
         $crawler = $this->createTestCrawler()->filterXPath('//ul[1]/li');
         $nodes = $crawler->reduce(function ($node, $i) {
-            return $i == 1 ? false : true;
+            return $i != 1;
```
Comment (WouterJ) [inline src/Symfony/Component/DomCrawler/Tests/CrawlerTest.php]: Symfony always uses `!==` were possible

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php [hunk 13]
```diff
@@ -663,14 +663,16 @@ public function remove($optionNames)
...
+            unset(
+                $this->defined[$option],
+                $this->defaults[$option],
+                $this->required[$option],
+                $this->resolved[$option],
+                $this->lazy[$option],
+                $this->normalizers[$option],
+                $this->allowedTypes[$option],
+                $this->allowedValues[$option]
+            );
```
Comment (WouterJ) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: this one too :)

File: src/Symfony/Component/Security/Acl/Dbal/AclProvider.php (hunk 11 above)
Comment (WouterJ) [inline src/Symfony/Component/Security/Acl/Dbal/AclProvider.php]: -1 on that

File: src/Symfony/Component/DomCrawler/Tests/CrawlerTest.php (hunk 12 above)
Comment (xabbuh) [inline src/Symfony/Component/DomCrawler/Tests/CrawlerTest.php]: But in this case, it didn't use a strict comparison before. Changing this behaviour now could lead to regressions.

File: src/Symfony/Component/Security/Acl/Dbal/AclProvider.php (hunk 9 above)
Comment (kalessil) [inline src/Symfony/Component/Security/Acl/Dbal/AclProvider.php]: @xabbuh : it's more readable actually, but maintainers have to decide if this stays or needs to be reverted =)

File: src/Symfony/Component/DomCrawler/Tests/CrawlerTest.php (hunk 12 above)
Comment (WouterJ) [inline src/Symfony/Component/DomCrawler/Tests/CrawlerTest.php]: as it's a test, it wouldn't hurt that much (and since `$i` is the loop variable, it will always be an integer)

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php (hunk 13 above)
Comment (kalessil) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: Emm, are you sure - it will be completely unreadable?

File: src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php (hunk 7 above)
Comment (kalessil) [inline src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php]: I'll check

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php (hunk 13 above)
Comment (javiereguiluz) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: @WouterJ in this case, a one-liner would look ugly, don't you think?

File: src/Symfony/Component/Security/Acl/Dbal/AclProvider.php (hunk 11 above)
Comment (javiereguiluz) [inline src/Symfony/Component/Security/Acl/Dbal/AclProvider.php]: The `int` prefix should be removed because Symfony doesn't follow this practice.

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php (hunk 13 above)
Comment (dupuchba) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: Where is the limit ?!

File: src/Symfony/Component/Security/Acl/Dbal/AclProvider.php (hunk 11 above)
Comment (kalessil) [inline src/Symfony/Component/Security/Acl/Dbal/AclProvider.php]: Ok, will drop prefixes int/str.

File: src/Symfony/Component/DomCrawler/Tests/CrawlerTest.php (hunk 12 above)
Comment (kalessil) [inline src/Symfony/Component/DomCrawler/Tests/CrawlerTest.php]: Will replace with !==

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php (hunk 13 above)
Comment (javiereguiluz) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: @dupuchba the limit would be the usual line length used in the rest of the codebase.

File: src/Symfony/Component/Security/Acl/Dbal/AclProvider.php [hunk 14]
```diff
@@ -192,7 +193,8 @@ public function findAcls(array $oids, array $sids = array())
                         $this->cache->putInCache($loadedAcl);
                     }
 
-                    if (isset($oidLookup[$loadedOid->getIdentifier().$loadedOid->getType()])) {
+                    $strOidLookupKey = $loadedOid->getIdentifier() . $loadedOid->getType();
+                    if (isset($oidLookup[$strOidLookupKey])) {
```
Comment (javiereguiluz) [inline src/Symfony/Component/Security/Acl/Dbal/AclProvider.php]: All the variable type prefixes (`$int...`, `$str...`, etc.) should be removed.

File: src/Symfony/Component/Security/Acl/Dbal/AclProvider.php (hunk 11 above)
Comment (javiereguiluz) [inline src/Symfony/Component/Security/Acl/Dbal/AclProvider.php]: @kalessil thanks! By the way, this is a nice PR :)

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php (hunk 13 above)
Comment (WouterJ) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: It would look ugly, as would many oneline code usages. However, Symfony CS specifies that there should be no maximum line length and oneline is prefered.

**Comment by kalessil** [comment]
@fabpot, @javiereguiluz I have to split PR into 2 parts:
- 2.3 : to include part not conflicted during rebase
- 2.7 : to include conflicted part, or maybe it will be other proposal here?

**Comment by kalessil** [comment]
Travis build failed during environment setup, nothing related to my changes

$ if [ "$TRAVIS_PHP_VERSION" != "hhvm-nightly" ]; then pecl install -f memcached-2.1.0; fi;
No releases available for package "pecl.php.net/memcached"
install failed

File: src/Symfony/Component/Security/Acl/Dbal/AclProvider.php (hunk 11 above)
Comment (kalessil) [inline src/Symfony/Component/Security/Acl/Dbal/AclProvider.php]: @javiereguiluz thanks :)

**Comment by kalessil** [comment]
PR split into <a href="https://github.com/symfony/symfony/pull/13862">PR 13862</a> and <a href="https://github.com/symfony/symfony/pull/13863">PR 13863</a>. Closing this one.
